### PR TITLE
Bit plane encoder

### LINKED ISSUE #16: CCSDS image data compression module
Since 3 of 5 OreSat Linux cards are camera cards, we should have a common image compression module.

CCSDS has a [Image Data Compression Blue Book](https://public.ccsds.org/Pubs/122x0b2.pdf)

It would be great to find a C, C++, or Rust library that implement this standard, that could be binded to Python and include in this framework (a pure Python module may be too slow).

### PR DESCRIPTION
**Pull request by kk60503**
Still working on binding

### CODE REVIEW

File: olaf/bpe/bpe.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+
+PYBIND11_MODULE(bpe, m)
+{
+    //docstring
+    m.doc() = "pybind11 bpe plugin";    
+        
+        py::class_<BITSTREAM>(m, "Bitstream")
+            .def(py::init<unsigned long, unsigned long, unsigned int, 
+                                            unsigned int, unsigned char>());
```
Comment (ryanpdx) [inline olaf/bpe/bpe.cpp]: This should be:
```
py::class_<BitStream>(m, "Bitstream")
        .def(py::init<>())
        .def_readonly("SegBitCounter", &BitStream::SegBitCounter)
        .def_readonly("TotalBitCounter", &BitStream::TotalBitCounter)
        .def_readonly("ByteBitCounter", &BitStream::ByteBuffer_4Bytes)
        .def_readonly("CodeWordAlighmentBits", &BitStream::CodeWordAlighmentBits)
        .def_readonly("CodeWord_Length", &BitStream::CodeWord_Length)
        .def_readonly("F_Bits", &BitStream::F_Bits);
```

File: olaf/bpe/bpe.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
+#include <Python.h>
+#include <structmember.h>
+#include <iostream>
+#include "pybind11/pybind11.h"
```
Comment (ryanpdx) [inline olaf/bpe/bpe.cpp]: This is a global include, so use "<" and ">"

File: olaf/bpe/bpe.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
+#include <Python.h>
+#include <structmember.h>
+#include <iostream>
```
Comment (ryanpdx) [inline olaf/bpe/bpe.cpp]: Not needed

File: olaf/bpe/bpe.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+#include "pybind11/pybind11.h"
+#include <pybind11/pytypes.h>
+#include <pybind11/numpy.h>
+#include <string>
```
Comment (ryanpdx) [inline olaf/bpe/bpe.cpp]: Not needed

File: olaf/bpe/bpe.cpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+#include "source/main_pybind.h"
```
Comment (ryanpdx) [inline olaf/bpe/bpe.cpp]: Shouldn't this be `"source/globals.h"`?

File: olaf/bpe/bpe.cpp [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+#include <structmember.h>
+#include <iostream>
+#include "pybind11/pybind11.h"
+#include <pybind11/pytypes.h>
+#include <pybind11/numpy.h>
+#include <string>
+#include "source/main_pybind.h" 
+
+namespace py = pybind11;
+using namespace pybind11::literals;
```
Comment (ryanpdx) [inline olaf/bpe/bpe.cpp]: Is this needed?

File: olaf/bpe/makefile [hunk 7]
```diff
@@ -0,0 +1,79 @@
+#Makefile for binding bpe C code to Python
```
Comment (ryanpdx) [inline olaf/bpe/makefile]: Should be Makefile not makefile.

File: olaf/bpe/source/README.rtf [hunk 8]
```diff
@@ -0,0 +1,7 @@
+Kyle Klein, Ryan Medick for OreSat - August 2023
+kleinky@pdx.edu - oresat@pdx.edu
+
+This program has been adapted from the University of Nebraska-Lincoln's 
+implementation of the CCSDS 122.0-B-2 Recommended Standard image compression algorithm.
+
+
```
Comment (ryanpdx) [inline olaf/bpe/source/README.rtf]: Remove my name and add a link to the website.

File: olaf/bpe/source/makefile [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# CCSDS image compression standard makefile for usage compiling as C code
```
Comment (ryanpdx) [inline olaf/bpe/source/makefile]: Makefile not makefile

File: olaf/bpe/source/UNL_readme_kielymods.rtf [hunk 10]
```diff
@@ -0,0 +1,424 @@
+{\rtf1\mac\ansicpg10000\cocoartf824\cocoasubrtf480
```
Comment (ryanpdx) [inline olaf/bpe/source/UNL_readme_kielymods.rtf]: Use the txt version, not rtf one

Comment (ryanpdx) [reply olaf/bpe/source/UNL_readme_kielymods.rtf]: The pdf is not needed.

File: olaf/bpe/source/README.rtf (hunk 8 above)
Comment (ryanpdx) [reply olaf/bpe/source/README.rtf]: Use md not rtf

File: olaf/bpe/source/pixel_check.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
+# pixel_check.py: a simple program for comparing/checking each pixel in identical images
```
Comment (ryanpdx) [inline olaf/bpe/source/pixel_check.py]: Remove this file

File: olaf/bpe/source/main_pybind.h [hunk 12]
```diff
@@ -0,0 +1,330 @@
+// Header file including global.h plus binding declarations
```
Comment (ryanpdx) [inline olaf/bpe/source/main_pybind.h]: Why does this exist, isn't it just `globals.h`?

File: olaf/bpe/makefile [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+#CXX = g++
+
+
+PYLIB_SRC = .
+
+BPE_SRC = source
+
+LINK_TARGET = bpe.so
+
+CPPFLAGS = \
```
Comment (ryanpdx) [inline olaf/bpe/makefile]: Rename to `INC_DIRS`

File: olaf/bpe/source/cfc.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
+# A simple program to convert raw cfc images to viewable file types
```
Comment (ryanpdx) [inline olaf/bpe/source/cfc.py]: Remove this file

File: olaf/bpe/source/errorhandle.c [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+"Error code 15: Pattern Coding Error." ,/*Coding Parameters  BPE_PATTERNING_CODING_ERROR 15 */
+"Error code 16: Rice Coding Error.", /*Coding Parameters  BPE_RICING_CODING_ERROR 16 */
+"Error code 17: Stage Coding Error." /*Coding Parameters  BPE_STAGE_CODING_ERROR 17 */};
+
+
+void DebugInfo(char *m)
+{
+#ifdef DEBUG
+	fprintf(stderr, m);
+#endif
```
Comment (ryanpdx) [inline olaf/bpe/source/errorhandle.c]: This give compile errors

```
source/errorhandle.c: In function ‘DebugInfo’:
source/errorhandle.c:48:9: error: format not a string literal and no format arguments [-Werror=format-security]
   48 |         fprintf(stderr, m);
      |         ^~~~~~~
```

Comment (ryanpdx) [reply olaf/bpe/source/errorhandle.c]: Commenting out #define DEBUG fixes it

File: olaf/bpe/makefile [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+		 $(shell python3 -m pybind11 --includes)
+
+CXXFLAGS = \
+		 -std=c++11 -g -Wall $(CPPFLAGS) -fPIC \
+		 $(shell python3-config --includes --cflags) \
+		 $(shell python3 -m pybind11 --includes)
+
+LDFLAGS = -lm $(shell python3-config --ldflags) \
+		  -undefined dynamic_lookup \
+		  #$(shell python3 -m pybind11 --includes)
```
Comment (ryanpdx) [inline olaf/bpe/makefile]: Remove commented out line

File: olaf/bpe/makefile [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+		 $(shell python3-config --includes --cflags) \
+		 $(shell python3 -m pybind11 --includes)
+
+CXXFLAGS = \
+		 -std=c++11 -g -Wall $(CPPFLAGS) -fPIC \
+		 $(shell python3-config --includes --cflags) \
+		 $(shell python3 -m pybind11 --includes)
+
+LDFLAGS = -lm $(shell python3-config --ldflags) \
+		  -undefined dynamic_lookup \
```
Comment (ryanpdx) [inline olaf/bpe/makefile]: Code won't compile with the `-undefined dyanmic_lookup` flag on my system. If it is removed, it compiles just fine.

**Review by ryanpdx — COMMENTED** [review]
Looks good, some clean up and changes to binding code needed.

File: olaf/bpe/testing/test_bpe.py [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+    def tearDown(self):
+        if os.path.isfile(self._compr_file):
+            os.remove(self._compr_file)
+        if os.path.isfile(self._outfile):
+            os.remove(self._outfile)
+        if os.path.isfile(self._png_infile):
+            os.remove(self._png_infile)
+        if os.path.isfile(self._png_outfile):
+            os.remove(self._png_outfile)
+
```
Comment (ryanpdx) [inline olaf/bpe/testing/test_bpe.py]: Remove, main is not required for unit tests

File: olaf/bpe/testing/test_bpe.py [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+from pixel_check import pixel_check
+
+
+class Test_bpe(unittest.TestCase):
+    def setUp(self):
+                            ## BEGIN user defined specs
+    # min encode params = (name of file to be encoded, name of newly encoded file,
+    #                                   width of image, height of image)
+    # min decode params = (name of encoded file to decode, name for the outfile)
+
```
Comment (ryanpdx) [inline olaf/bpe/testing/test_bpe.py]: Use """ comment sytle for all docstrings

File: olaf/bpe/testing/test_bpe.py [hunk 20]
```diff
@@ -0,0 +1,88 @@
+# testing suite for bpe python module
```
Comment (ryanpdx) [inline olaf/bpe/testing/test_bpe.py]: change to """ comment style

File: olaf/bpe/main_pybind.cpp [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+     fprintf(stderr, "eg 2: bpe -d codes -o ss.img \n");                        
+     fprintf(stderr, "**********  Author: Hongqiang Wang  *****************\n");
+     fprintf(stderr, "**********  Department of Electrical Engineering ****\n");
+     fprintf(stderr, "**********  University of Nebraska -Lincoln  ********\n");
+     fprintf(stderr, "**********  March 9, 2008   *************************\n");
+     fprintf(stderr, "/****************************************************\n");
+     fprintf(stderr, "/*****  Altered Sept 2023 for OreSat/PSAS  **********\n");
+     fprintf(stderr, "/****************************************************\n");
+    return;                                                                     
+}
```
Comment (ryanpdx) [inline olaf/bpe/main_pybind.cpp:37]: Remove

File: olaf/bpe/testing/pixel_check.py [hunk 22]
```diff
@@ -0,0 +1,38 @@
+# pixel_check.py: 
+#  a simple program for comparing/checking each pixel in identical images
+# **currently has only been tested on PNG image type
+# adapted from:
+#  https://stackoverflow.com/questions/62285511/use-numpy-to-quickly-iterate-over-pixels
+# (program utilizes PIL)
```
Comment (ryanpdx) [inline olaf/bpe/testing/pixel_check.py]: Use block comment style

File: olaf/bpe/testing/pixel_check.py [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+            if pixel_val1 != pixel_val2:
+                print(f"Found pixel {pixel_val2} at {i, j} does not match")
+
+    print('scan complete')
+
+# if calling from command line: python pixel_check.py <file1> <file2>
+if __name__ == '__main__':
+    file1 = sys.argv[1]
+    file2 = sys.argv[2]
+    pixel_check(file1, file2)
```
Comment (ryanpdx) [inline olaf/bpe/testing/pixel_check.py]: Use argparser

**Review by ryanpdx — CHANGES_REQUESTED** [review]
- Remove all `.pyc` files
- move `olaf/bpe/testing/` contents  to the `tests/bpe/`
- move all images out of the `olaf/` dir
- What are the `GHplaceholder` files

**Comment by ryanpdx** [comment]
If you can, please try to squash some of your commits and update your comments to better reflect what changed in the commits. 

Also run `black` and `isort` on your `.py` files.

File: olaf/bpe/testing/test_bpe.py (hunk 18 above)
Comment (kk60503) [reply olaf/bpe/testing/test_bpe.py]: Sorry Ryan, did you want me to remove lines 86-down at this part?

**Pull request closed without merging**